test/evp_pkey_provided_test.c

```diff
@@ -443,6 +443,110 @@ static int test_fromdata_rsa(void)
     return ret;
 }
 
+static int test_fromdata_rsa_derive_from_pq(void)
+{
+    OSSL_PARAM_BLD *bld = NULL;
+    BIGNUM *n = NULL, *e = NULL, *d = NULL, *p = NULL, *q = NULL;
+    OSSL_PARAM *fromdata_params = NULL;
+    EVP_PKEY_CTX *ctx = NULL, *key_ctx = NULL;
+    EVP_PKEY *pk = NULL, *copy_pk = NULL, *dup_pk = NULL;
+    int ret = 0;
+    /*
+     * 512-bit RSA key, extracted from this command,
+     * openssl genrsa 512 | openssl rsa -text
+     * Note: When generating a key with EVP_PKEY_fromdata, and using
+     * crt derivation, openssl requires a minimum of 512 bits of n data, and 2048 bits in the FIPS case
+     */
+    static unsigned char n_data[] = {0x00,0xb6,0x5a,0xb5,0x70,0xc3,0x34,0x13,0x20,0xaa,0xec,0x30,0x34,0x33,0x75,
+    0xe4,0xb4,0xab,0xce,0x3f,0x95,0x2d,0xd7,0xae,0x89,0x65,0x86,0x6e,0xfd,0x27,
+    0xf5,0x94,0x90,0x04,0xcf,0x00,0x06,0xfd,0xc5,0x1c,0x6b,0xbf,0x61,0xc7,0xe1,
+    0x2d,0x9e,0xb6,0x49,0x60,0x8f,0x87,0x03,0x85,0x33,0xdc,0xe2,0xe3,0x89,0x76,
+    0xb8,0xa9,0x71,0x41,0xe3};
+
+    static unsigned char e_data[] = {0x01, 0x00, 0x01};
+
+    static unsigned char d_data[] = {0x78,0xe6,0x9d,0x23,0x5d,0x18,0x9a,0x91,0xef,0x20,0xd7,0x38,0xd4,0xe1,0x96,
+    0x37,0xbf,0x2c,0xe8,0x75,0x04,0xc6,0xd9,0xfe,0x58,0x64,0x90,0x13,0xa2,0x39,
+    0xad,0x3b,0xcc,0xa2,0xd7,0x23,0xd9,0xab,0xe5,0xdc,0x22,0x4e,0xf4,0xc8,0xb3,
+    0x4d,0xcb,0x72,0x6a,0xd1,0x02,0xa5,0x12,0x56,0x53,0x76,0x3f,0x9d,0xe2,0x0b,
+    0x7e,0x5f,0x51,0x31};
+
+    static unsigned char p_data[] = {0x00,0xd9,0x2e,0x3d,0x42,0x27,0x86,0xd9,0x82,0x1b,0x53,0x6f,0xf5,0xee,0x69,
+    0x16,0x1f,0x52,0xe4,0x3d,0xe2,0xa7,0x39,0x76,0x7b,0x0c,0xe4,0x36,0x7b,0x6f,
+    0xfe,0x2c,0x2b};
+
+    static unsigned char q_data[] = {0x00,0xd6,0xf2,0xe2,0xee,0xf8,0x31,0x50,0x6f,0x14,0xb1,0x62,0x3c,0xc0,0x88,
+    0xfe,0x9c,0xde,0x82,0x6b,0x0b,0xe0,0x06,0xad,0xf3,0x72,0x94,0x88,0x44,0xcf,
+    0x7a,0x0d,0x29};
+
+    if (!TEST_ptr(bld = OSSL_PARAM_BLD_new())
+        || !TEST_ptr(n = BN_bin2bn(n_data, sizeof(n_data), NULL))
+        || !TEST_ptr(e = BN_bin2bn(e_data, sizeof(e_data), NULL))
+        || !TEST_ptr(d = BN_bin2bn(d_data, sizeof(d_data), NULL))
+        || !TEST_ptr(p = BN_bin2bn(p_data, sizeof(p_data), NULL))
+        || !TEST_ptr(q = BN_bin2bn(q_data, sizeof(q_data), NULL))
+        || !TEST_true(OSSL_PARAM_BLD_push_BN(bld, OSSL_PKEY_PARAM_RSA_N, n))
+        || !TEST_true(OSSL_PARAM_BLD_push_BN(bld, OSSL_PKEY_PARAM_RSA_E, e))
+        || !TEST_true(OSSL_PARAM_BLD_push_BN(bld, OSSL_PKEY_PARAM_RSA_D, d))
+        || !TEST_true(OSSL_PARAM_BLD_push_BN(bld, OSSL_PKEY_PARAM_RSA_FACTOR1, p))
+        || !TEST_true(OSSL_PARAM_BLD_push_BN(bld, OSSL_PKEY_PARAM_RSA_FACTOR2, q))
+        || !TEST_true(OSSL_PARAM_BLD_push_int(bld, OSSL_PKEY_PARAM_RSA_DERIVE_PQ, 1))
+        || !TEST_ptr(fromdata_params = OSSL_PARAM_BLD_to_param(bld))
+        || !TEST_ptr(ctx = EVP_PKEY_CTX_new_from_name(NULL, "RSA", NULL))
+        || !TEST_int_eq(EVP_PKEY_fromdata_init(ctx), 1)
```

Comment:
Some thoughts on your test...
I think you should have 2 known answer tests by grabbing the data from RSA keygen calls.
1 that does more than 2 multiprimes, and one that uses the sp80056a path.
It should extract out the dp,dQ, qinv and then use these as expected values in the from data calls.